src/rust/engine/options/src/config.rs

```diff
@@ -1,18 +1,94 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::mem;
 use std::path::Path;
 
+use regex::Regex;
 use toml::value::Table;
 use toml::Value;
 
 use super::id::{NameTransform, OptionId};
 use super::parse::parse_string_list;
 use super::{ListEdit, ListEditAction, OptionsSource};
 
+type InterpolationMap = HashMap<String, String>;
+
+pub(crate) fn interpolate_string(
```

Comment:
Is `pub(crate)` the right way to make this function available to the test? Is there a more restrictive way to do that without exposing to the entire crate?

Reply:
The usual (in the broader Rust ecosystem) way to test private functions like this is having the tests as submodules: a 'private' function like `foo::bar::some_func` visible down the tree, e.g. `foo::bar::baz::another_func` or `foo:bar::tests::test_some_func` can call it.

Pants often seems to put tests in sibling modules, rather than submodules, which I think stops this working quite as naturally. Given that restriction a `pub(crate)` (or maybe `pub(super)`, although that's equivalent for this particular module, since the nesting means `super` == the crate root).

One way to do this without the `pub(crate)` would be to change to that structure:

- move `config_tests.rs` to `config/tests.rs` and move the `#[cfg(tests)] mod config_tests;` in `lib.rs` to `#[cfg(tests)] mod tests;` in this file, to make it a submodule.
- (One could then also move to this file to `config/mod.rs`, which is still exposed as `options::config`, but that's not _necessary_.)

Reply:
Yeah, I don't want to start changing things up too much here and now. But it's something to think about as a policy.